adapters/repos/db/shard.go

```diff
@@ -58,10 +64,24 @@ type Shard struct {
 	versioner        *shardVersioner
 	diskScanState    *diskScanState
 
+	numActiveBatches    int
+	activeBatchesLock   sync.Mutex
+	jobQueueCh          chan job
```

Comment:
The new approach look to me like every shard having its own queue of jobs. Each batch request adds objects to the queue. The workers fetch objects from the queue and process them.

This approach has the advantage that it uses a fixed number of go-routines and hence uses far less go-routines than the current implementation.

However the new approach may lead to small batches taking too much time to complete or even timing out because jobs from previous batches haven’t finished. This case will be more apparent when we have batches of different size mixed together. This is not an issue in the current implementation because batch requests are processed independently

The problem is inherent to the approach itself because it doesn’t take into account the batch size during insertion.

Reply:
discussed in slack :)